### DISCUSSION TITLE
Setting Exercise Time Issue

### DISCUSSION THREAD

**Discussion by RubyBailey** [jgyates/genmon#1215] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Version

Genmon Version V1.19.04

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

The issue is that I am unable to set certain Exercise Date / Time through Genmon.

Preamble and explanation: The generator exercise day and time were set on the generator to Thursday, 12.08. It has been running that way for quite some time. I changed the exercise day to Wednesday using Genmon. It was easier that walking to the generator on a rainy day. That worked flawlessly. After some discussion it was decided that remaining on Thursday would be better so I tried to change the day back to Thursday using Genmon and found that I could not set the day back to the previous weekday. I then did some testing and found the following:

1. It is possible to set the day to any weekday except Thursday.

2. The time set can be 1 to 2 minutes earlier than the time selected in the Generator Exercise Time selection settings. See below copied from Maintenance screen. Exercise time was set by 'Set Exercise Time' 12.38, but actually set on the generator to 12:36. This is not necessarily a big issue just something noted during the testing.

```    Exercise :
        Exercise Time : Weekly Wednesday 12:36 Quiet Mode On

    Service :
        Air Filter Service Due : 112 hrs or 11/06/2025
        Oil and Oil Filter Service Due : 162 hrs or 11/06/2026
        Spark Plug Service Due : 162 hrs or 11/06/2026
        Battery Service Due : 162 hrs or 11/06/2026
        Total Run Hours : 1026.00 h
        Hardware Version : V1.02
        Firmware Version : V1.42

Generator Exercise Time:
Wednesday  ,12 :38

```

### Additional information
From mymodbus.log 
2024-12-05 22:38:55,264 : Error in GetCRC: byte must be in range(0, 256) : mymodbus.py:1067
2024-12-05 22:38:55,265 : Packet = [0x9d,0x10,0x00,0x04,0x00,0x01,0x02,0x-2,0xd5]
2024-12-05 22:38:55,265 : Error in SendPacketAsMaster: byte must be in range(0, 256) : mymodbus.py:843
2024-12-05 22:38:55,266 : Packet = [0x9d,0x10,0x00,0x04,0x00,0x01,0x02,0x-2,0xd5]
2024-12-05 22:38:58,294 : Error: timeout receiving slave packet for register 0004 Buffer: 0, sequence 36365246

From Genmon Maintenance page:
Maintenance :
    Model : Generic Air Cooled
    Generator Serial Number : 0007168870
    Controller Detected : Nexus, Air Cooled
    Nominal RPM : 3600
    Rated kW : 20 kW
    Nominal Frequency : 60 Hz
    Fuel Type : Propane
    Generator Phase : 1
    Engine Displacement : 999 cc
    Controller Settings :
        Calibrate Volts : 1009
        Nominal Line Voltage : 240 V
        Rated Max Power : 20 kW
    Exercise :
        Exercise Time : Weekly Wednesday 13:02 Quiet Mode On
    Service :
        Air Filter Service Due : 112 hrs or 11/06/2025
        Oil and Oil Filter Service Due : 162 hrs or 11/06/2026
        Spark Plug Service Due : 162 hrs or 11/06/2026
        Battery Service Due : 162 hrs or 11/06/2026
        Total Run Hours : 1026.00 h
        Hardware Version : V1.02
        Firmware Version : V1.42

From Genmon Monitor page:
Monitor :
    Generator Monitor Stats :
        Monitor Health : OK
        Controller : Nexus, Air Cooled
        Run time : Generator Monitor running for 13 days, 1:29:42.
        Generator Monitor Version : V1.19.04
        Update Available : No
    Communication Stats :
        Packet Count : M: 36289280, S: 36289279
        CRC Errors : 0
        CRC Percent Errors : 0.00%
        Timeout Errors : 16
        Timeout Percent Errors : 0.00%
        Modbus Exceptions : 0
        Validation Errors : 0
        Sync Errors : 0
        Invalid Data : 0
        Discarded Bytes : 0
        Comm Restarts : 0
        Packets Per Second : 64.31
        Average Transaction Time : 0.0307 sec
        Modbus Transport : Serial
        Serial Data Rate : 9600
    Platform Stats :
        CPU Temperature : 33.20 C
        Pi Model : Raspberry Pi 3 Model B Plus Rev 1.3
        Pi CPU Frequency Throttling : Has occurred. Throttling Active.
        Pi ARM Frequency Cap : OK
        Pi Undervoltage : Has occurred. Undervoltage Detected.
        CPU Utilization : 4.0%
        OS Name : Debian GNU/Linux
        OS Version : 12 (bookworm)
        System Uptime : 22 days, 5:51:26
        Network Interface Used : wlan0
        WLAN Signal Level : -78 dBm
        WLAN Signal Quality : 32/70
        WLAN Signal Noise : -256 dBm
        WLAN ESSID : [SSID removed]
        System Time : Thursday December 5, 2024 21:58:43

**Comment by jgyates** [comment, marked as answer]
There is a know limitation with nexus controllers when setting the exerciser time:

https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues#2-setting-nexus-exercise

Generac addressed some of these issues with setting the controller with firmware updates however it was not until the Evolution model that they fundamentally changed the way the exercise time and date was set via modbus.

Regarding the errors you posted in your mymodbus.log file. This is an issue with serial communications. It is hard to say what the issue is without more info (full logs) however you do appear to have a power issue as you pi looks like it has these errors:

Pi CPU Frequency Throttling : Has occurred. Throttling Active.
Pi Undervoltage : Has occurred. Undervoltage Detected.

These errors can be an indicator of a suboptimal power supply. Are you powering your pi from the molex connector or the battery? Powering from the molex can have some issues, particularly on Nexus models as it may not supply enough current. If you are powering form the battery then you may want to check your power and ground connections.

**Reply by RubyBailey** [reply]
Thanks for the quick response. I had to read the known issue a couple of times to understand it and yes that was / is the problem. I can work within those limitations. 
I will follow up on the throttling / undervoltage issue. The pi is powered from the battery with a 5V/2.5A rated supply. Obviously I have an issue there somewhere. Thanks for noting that.
